proposals/0387-bls-pubkey-management-in-vote-account.md

```diff
@@ -0,0 +1,206 @@
+the current ed25519 public keys in vote accounts either. We have to add a BLS
+public key into each vote account and make that a requirement a little before
+Alpenglow launches.
+
+## Dependencies
+
+- Alpenglow is specified in [SIMD 326](https://github.com/solana-foundation/solana-improvement-documents/pull/326)
+
+- VoteStateV4 is specified in [SIMD 185](https://github.com/solana-foundation/solana-improvement-documents/pull/185)
+and it adds an optional BLS public key field
+
+- Requiring BLS public key for Alpenglow is specified in [SIMD 357](https://github.com/solana-foundation/solana-improvement-documents/pull/357)
+
+## New Terminology
+
+- **BLS public key**: The public key used in BLS signatures. It will be
+compressed and stored as a 48 bytes vector in vote account.
+
+## Detailed Design
+
+BLS keypairs can be generated randomly like ed25519 keypairs. But to save the
+users some trouble on keypair management, we chose to derive their BLS keypair
+used in Alpenglow votes based on their ed25519 vote keypair. In other words,
+with an existing ed25519 vote keypair, the users can safely regenerate the
+associated BLS keypair on demand.
+
+### User Operations Support
+
+Since a user’s BLS keypair always changes with the vote authority keypair, we
+list three different scenarios below. The following are all dependent on the
+feature associated with this SIMD being rolled out, the operations will not
+change before the feature is active.
+
+1. Creating a new vote account
+
+When users create a new vote account, the correct BLS public key will be
+automatically added to the new vote account. The BLS public key is derived
+from the vote authority keypair (if authorized voter is not specified, identity
+keypair is used).
+
+2. Changing vote authority keypair on existing vote account
+
+When users update the vote authority keypair, the correct BLS public key will
+be automatically updated in the vote account. The BLS public key rotation
+happens at the same time when a new vote authority switch happens.
+
+3. Updating missing BLS public key on existing vote account
+
+This is a temporary work-around before all existing staked vote accounts have
+proper BLS public keys specified. We will add `update-bls-pubkey` command so it
+will use the given vote authority keypair to generate and update the BLS public
+key in the vote account. When everyone has a proper BLS public key in their
+vote accounts this command can be removed.
+
+### Changes to vote program
+
+Whenever a new BLS public key is being updated in the vote account, we need
+to perform BLS verification on its validity, see "Security Considerations"
+for details. We plan to implement this by calling BLS library from the vote
+program, see "Alternatives Considered" for comparison with other solutions.
+
+Since BLS verification is expensive (in the order of ~1 millisecond), these
+operations changing BLS public key will need to have correct CU specified to
```

Comment:
Can you clarify that CU's must be consumed right before doing BLS signature verification? And cc @tao-stones for discussing the CU number given time for BLS verification.

Reply:
Yup, already asked Tao and he estimate 1.15ms running time translates to 34,500 CU. Updated that in the PR.